### PR TITLE
Task: Implement lazy loading

### PR DESCRIPTION
**Pull request by Onokaev**
## Overview
Implements #2228 

### Demo

Optional. Screenshots, `curl` examples, etc.
image
Chunks before lazy loading

image

Chunks after lazy loading

### Notes
Currently, Graph Explorer is loaded in a single bundle. With the app growing, this bundle is also increasing in size. The current bundle has a size a little over 1MB. This has an impact on the app load time as the whole bundle has to be loaded first for the app to become usable

Lazy loading splits the original bundle into smaller chunks so that what is needed on first-load is what is loaded initially. The lazily loaded components will be in their own chunks and will be loaded on demand

## Testing Instructions

* How to test this PR
Load the app on your browser
Click 'Request headers' tab
Notice that it is loaded on demand
Click on other components (Modify Permissions, Access token, Response headers, Resources etc.)

### CODE REVIEW

File: src/app/views/query-response/headers/index.ts [hunk 1]
```diff
@@ -1,2 +1,2 @@
 import ResponseHeaders from './ResponseHeaders';
-export { ResponseHeaders };
+export default ResponseHeaders;
```
Comment (thewahome) [inline src/app/views/query-response/headers/index.ts:2]: could `export { ResponseHeaders as default };`  also work?

File: src/app/views/query-response/pivot-items/pivot-items.tsx [hunk 2]
```diff
@@ -1,5 +1,5 @@
-import { getTheme, IPivotItemProps, ITheme, PivotItem } from '@fluentui/react';
-import React from 'react';
+import { getTheme, IPivotItemProps, ITheme, PivotItem, Spinner, SpinnerSize } from '@fluentui/react';
```
Comment (thewahome) [inline src/app/views/query-response/pivot-items/pivot-items.tsx:1]: These are repeated heavily, we can have a `SuspenseLoader` component that we can use to wrap all the suspense content

**Review by thewahome — COMMENTED** [review]
Let's have the PR body explain Please also mention in the PR description the gains expected. 
Are we aiming to reduce bundle size? How much does the work in this PR save us?

IMHO it could be that the network calls we are making on the first load are expensive. And that it could be unrelated to lazy loading the different tabs.

We can also analyse whether GE with the Dev portal chrome around it gives us the same load time as GE without the portal chrome. Then see whether it's the extra html added during production that affects the initial load time

File: config/webpack.config.js [hunk 3]
```diff
@@ -414,7 +414,7 @@ module.exports = function (webpackEnv) {
     },
     plugins: [
       new webpack.optimize.LimitChunkCountPlugin({
-        maxChunks: 1
+        maxChunks: 100
```
[the remark this answers is not in the document]
Comment (Onokaev) [reply config/webpack.config.js:417]: Components that are lazy loaded are in their own chunks. The benefit of this is that the initial chunk that is loaded (with the components that a user sees for the first time) is small. So, it loads much faster. 
When you click on the Modify Permissions tab, it's chunk is loaded.

File: src/app/views/query-response/headers/index.ts (hunk 1 above)
Comment (Onokaev) [reply src/app/views/query-response/headers/index.ts:2]: Yes. Let me change this

**Review by thewahome — CHANGES_REQUESTED** [review]
Since some components are not immediately with the client when the page loads, they may fail to get downloaded eventually when needed due to various reasons. It would make sense to add an `ErrorBoundary` wrapped around the `SuspenseLoader`

**Review by thewahome — COMMENTED** [review]
We are going to add a lot of SuspenseLoaders everywhere. 

It's probably possible for us to have a component registry where we register the suspended components and import them in the required files without adding the suspense and error boundaries.

If successful, it could mean we would be changing only where the components are imported from and not how they are imported. 

e.g.
`import { Permission } from 'component-registry';` instead of `const Permission = lazy(() => import('../../query-runner/request/permissions'))`

**Comment by calebkiage** [comment]
You could also explore only lazy loading non-essential modules, then chunking and loading all essential modules. Multiple smaller files might load faster than a single large file due to parallel network downloads. However, the chunk size also plays a part. If the chunks are too small, you waste time in creating the connection before downloading.

File: src/app/views/query-response/adaptive-cards/AdaptiveCard.tsx [hunk 4]
```diff
@@ -126,7 +126,7 @@ const AdaptiveCard = (props: any) => {
       }
       return (
         <Pivot className='adaptive-pivot'
-          onLinkClick={(pivotItem) => onPivotItemClick(sampleQuery, pivotItem)}
+          onLinkClick={(pivotItem: any) => onPivotItemClick(sampleQuery, pivotItem)}
```
Comment (gavinbarron) [inline src/app/views/query-response/adaptive-cards/AdaptiveCard.tsx:129]: We shouldn't use any unless we have no other choice
```suggestion
          onLinkClick={(pivotItem: PivotItem) => onPivotItemClick(sampleQuery, pivotItem)}
```

**Review by gavinbarron — CHANGES_REQUESTED** [review]

File: src/app/views/common/lazy-loader/component-registry/index.tsx [hunk 5]
```diff
@@ -0,0 +1,133 @@
+/* eslint-disable max-len */
+import { lazy } from 'react';
+import { IPermissionProps } from '../../../../../types/permissions';
+import { SuspenseLoader } from '../suspense-loader/SuspenseLoader';
+
+const LazyPermission = lazy(() =>
+/* webpackChunkName: "permissions" */ import('../../../query-runner/request/permissions'));
```
Comment (gavinbarron) [inline src/app/views/common/lazy-loader/component-registry/index.tsx:7]: The webpackChunkName magic comment needs to be inside the import() function call to ensure that the chunk actually gets the name you want it to

File: src/app/views/common/lazy-loader/component-registry/index.tsx [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+const LazyStatusMessages = lazy( () =>
+/* webpackChunkName: "statusMessages" */ import('../../../app-sections/StatusMessages'));
```
Comment (gavinbarron) [inline src/app/views/common/lazy-loader/component-registry/index.tsx:9]:
```suggestion
    import(/* webpackChunkName: "statusMessages" */ '../../../app-sections/StatusMessages'));
```

**Review by gavinbarron — APPROVED** [review]
LGTM

File: src/app/views/query-runner/request/Request.tsx [hunk 7]
```diff
@@ -74,7 +76,7 @@ const Request = (props: any) => {
         }}
       >
         <div style={containerStyle} id={'request-header-tab'}>
-          <RequestHeaders />
+          <RequstHeaders />
```
Comment (thewahome) [inline src/app/views/query-runner/request/Request.tsx:79]:
```suggestion
          <RequestHeaders />
```

File: src/app/views/query-runner/request/Request.tsx [hunk 8]
```diff
@@ -88,7 +90,7 @@ const Request = (props: any) => {
         }}
       >
         <div style={containerStyle} id={'permission-tab'}>
-          <Permissions />
+          <Permission />
```
Comment (thewahome) [inline src/app/views/query-runner/request/Request.tsx:93]:
```suggestion
          <Permissions />
```

File: src/app/views/query-runner/request/Request.tsx [hunk 9]
```diff
@@ -10,11 +14,9 @@ import { Mode } from '../../../../types/enums';
 import { setDimensions } from '../../../services/actions/dimensions-action-creator';
 import { translateMessage } from '../../../utils/translate-messages';
 import { convertPxToVh, convertVhToPx } from '../../common/dimensions/dimensions-adjustment';
-import { Auth } from './auth';
 import { RequestBody } from './body';
-import { RequestHeaders } from './headers';
-import { Permissions } from './permissions';
 import './request.scss';
+import { Permission, Auth, RequstHeaders } from '../../common/lazy-loader/component-registry';
```
Comment (thewahome) [inline src/app/views/query-runner/request/Request.tsx:19]:
```suggestion
import { Permission, Auth, RequestHeaders } from '../../common/lazy-loader/component-registry';
```

File: src/app/views/common/lazy-loader/component-registry/index.tsx [hunk 10]
```diff
@@ -0,0 +1,156 @@
...
+
+export const Auth = (props?: any) => {
+  return (
+    <SuspenseLoader>
+      <LazyAuth {...props} />
+    </SuspenseLoader>
+  )
+}
+
+export const RequstHeaders = (props?: any) => {
```
Comment (thewahome) [inline src/app/views/common/lazy-loader/component-registry/index.tsx:101]:
```suggestion
export const RequestHeaders = (props?: any) => {
```

**Review by thewahome — COMMENTED** [review]
Noticed a few typos

**Review by thewahome — APPROVED** [review]

**Comment by gavinbarron** [comment]
> Noticed a few typos

Nice spotting @thewahome 
I'd suggest that everyone uses the CSpell extension for VSCode, it really helps catch those kind of typos early.

**Comment by thewahome** [comment]
@gavinbarron I'll add it as a recommended extension

**Pull request merged**